### PR TITLE
[IMP] l10n_ph: new tax tags, sale withholding taxes

### PR DESCRIPTION
**Pull request by tong-odoo**
This commit improve below features:

- New tax tags QAPA, QAPB for purchase withholding taxes
- New tax tags SAWTA, SAWTB for sale withholding taxes
- Add sale withholding taxes
- Extract bir report exporter for l10n_ph_accounts to reuse

task-3700586

See https://github.com/odoo/enterprise/pull/67860

### CODE REVIEW

File: addons/l10n_ph/tests/common.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+try:
+    from openpyxl import load_workbook
+except ImportError:
+    load_workbook = None
+
+from odoo.addons.account.tests.common import AccountTestInvoicingCommon
+
+
+class TestPh(AccountTestInvoicingCommon):
```
Comment (vin-odoo) [inline addons/l10n_ph/tests/common.py:13]:
```suggestion
class TestPhCommon(AccountTestInvoicingCommon):
```

File: addons/l10n_ph/tests/test_bir_2307_generation.py [hunk 2]
```diff
@@ -64,15 +44,7 @@ def test_bir_2307_company(self):
             # Header
             0: ['Reporting_Month', 'Vendor_TIN', 'branchCode', 'companyName', 'surName', 'firstName', 'middleName', 'address', 'nature', 'ATC', 'income_payment', 'ewt_rate', 'tax_amount'],
             # Row
-            1: ['01/15/2020', '789456123', '789', 'Test Partner', '', '', '', '9 Super Street, Super City, Philippines', 'Test line', 'WC640', 100.0, -1.0, -1.0],
+            1: ['01/15/2020', '789456123', '789', 'Test Partner', '', '', '', '9 Super Street, Super City, Philippines', 'Test line', 'WC640', 100.0, 1.0, 1.0],
```
Comment (vin-odoo) [inline addons/l10n_ph/tests/test_bir_2307_generation.py:47]: Is this change on purpose?

File: addons/l10n_ph/wizard/generate_2307_wizard.py [hunk 3]
```diff
@@ -74,21 +24,12 @@ def action_generate(self):
...
-
-        for index, col_header in enumerate(COLUMN_HEADER_MAP.keys()):
-            worksheet.write(0, index, label=col_header)
-
-        self._write_rows(worksheet, self.moves_to_export)
-
-        workbook.save(file_data)
-        file_data.seek(0)
-        self.generate_xls_file = base64.b64encode(file_data.read())
+        bir_exporter = Bir2307Exporter(_('Form2307'), self.moves_to_export, file_format='xls')
```
Comment (vin-odoo) [inline addons/l10n_ph/wizard/generate_2307_wizard.py:27]: Do we really need to translate the file name?

**Review by vin-odoo — COMMENTED** [review]
Few comments/questions for later, no full review yet

File: addons/l10n_ph/tests/test_bir_2307_generation.py (hunk 2 above)
Comment (tong-odoo) [reply addons/l10n_ph/tests/test_bir_2307_generation.py:47]: Yes

File: addons/l10n_ph/wizard/generate_2307_wizard.py [hunk 4]
```diff
@@ -74,21 +24,12 @@ def action_generate(self):
...
-        for index, col_header in enumerate(COLUMN_HEADER_MAP.keys()):
-            worksheet.write(0, index, label=col_header)
-
-        self._write_rows(worksheet, self.moves_to_export)
-
-        workbook.save(file_data)
-        file_data.seek(0)
-        self.generate_xls_file = base64.b64encode(file_data.read())
+        bir_exporter = Bir2307Exporter('Form2307', self.moves_to_export, file_format='xls')
+        self.generate_xls_file = base64.b64encode(bir_exporter.export())
```
Comment (vin-odoo) [inline addons/l10n_ph/wizard/generate_2307_wizard.py:28]:
```suggestion
        self.xls_file = base64.b64encode(bir_exporter.export())
```
The name isn't very accurate

File: addons/l10n_ph/wizard/generate_2307_wizard.py [hunk 5]
```diff
@@ -74,21 +24,12 @@ def action_generate(self):
...
-
-        for index, col_header in enumerate(COLUMN_HEADER_MAP.keys()):
-            worksheet.write(0, index, label=col_header)
-
-        self._write_rows(worksheet, self.moves_to_export)
-
-        workbook.save(file_data)
-        file_data.seek(0)
-        self.generate_xls_file = base64.b64encode(file_data.read())
+        bir_exporter = Bir2307Exporter('Form2307', self.moves_to_export, file_format='xls')
```
Comment (vin-odoo) [inline addons/l10n_ph/wizard/generate_2307_wizard.py:27]: Will you ever need to use this class differently than just doing
- Create exporter
- Export
- Close exporter
?

If not, you could simply create a method in util that does all that at once and returns the file right away in one line.
self.xls_file = utils._export_bit_2307('Form2307', self.moves_to_export, file_format='xls')

This would make it a bit easier to use

File: addons/l10n_ph/wizard/generate_2307_wizard.py [hunk 6]
```diff
@@ -74,21 +24,12 @@ def action_generate(self):
...
+        self.generate_xls_file = base64.b64encode(bir_exporter.export())
+        bir_exporter.close()
 
         return {
             "type": "ir.actions.act_url",
             "target": "self",
-            "url": "/web/content?model=l10n_ph_2307.wizard&download=true&field=generate_xls_file&filename=Form_2307.xls&id={}".format(self.id),
+            "url": "/web/content?model=l10n_ph_2307.wizard&download=true&field=generate_xls_file&filename={}.xls&id={}".format(_('Form_2307'), self.id),
```
Comment (vin-odoo) [inline addons/l10n_ph/wizard/generate_2307_wizard.py:34]: You're still translating the name here, but I think it's not needed.
Unless we got feedback for that (but I wouldn't think so), better keep it simple

**Review by vin-odoo — COMMENTED** [review]
Not much to say, good job :smile:

File: addons/l10n_ph/utils.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            'company_name': partner.commercial_partner_id.name,
+            'first_name': partner.first_name or '',
+            'middle_name': partner.middle_name or '',
+            'last_name': partner.last_name or '',
+            'address': ', '.join([val for val in partner_address_info if val])
+        }
+        for invoice_line in move.invoice_line_ids.filtered(lambda line: line.display_type not in ('line_note', 'line_section')):
+            for tax in invoice_line.tax_ids.filtered(lambda x: x.l10n_ph_atc):
+                product_name = invoice_line.product_id.name or invoice_line.name
+                values['product_name'] = re.sub(r'[\(\)]', '', product_name) if product_name else ""
```
Comment (vin-odoo) [inline addons/l10n_ph/utils.py:66]:
```suggestion
                values['product_name'] = re.sub(r'[()]', '', product_name) if product_name else ""
```
same here

**Review by vin-odoo — CHANGES_REQUESTED** [review]
Besides that it LGTM

**Review by vin-odoo — APPROVED** [review]

**Comment by vin-odoo** [comment]
@robodoo r+

Good work, thanks!
